File: src/main/java/com/milaboratory/mixcr/cli/CommandBAM2fastq.java
package com.milaboratory.mixcr.cli;

import com.milaboratory.core.io.sequence.PairedRead;
import com.milaboratory.core.io.sequence.SequenceRead;
import com.milaboratory.core.io.sequence.SingleRead;
import com.milaboratory.core.io.sequence.fastq.PairedFastqWriter;
import com.milaboratory.core.io.sequence.fastq.SingleFastqWriter;
import com.milaboratory.mixcr.bam.BAMReader;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Command(name = "BAM2fastq", hidden = true,
        description = "Converts BAM/SAM file to paired/unpaired fastq files")
public class CommandBAM2fastq extends MiXCRCommand {
    @Option(names = {"-b", "--bam"}, description = "BAM files for conversion.", required = true)
    public String[] bamFiles;

    @Option(names = {"-r1"}, description = "File for first reads.", required = true)
    public String fastq1;

    @Option(names = {"-r2"}, description = "File for second reads.", required = true)
    public String fastq2;

    @Option(names = {"-u"}, description = "File for unpaired reads.", required = true)
    public String fastqUnpaired;

    @Option(names = {"--drop-non-vdj"},
            description = "Drop reads from bam file mapped on human chromosomes except with VDJ region (2, 7, 14, 22)")
    public boolean dropNonVDJ = false;

    @Option(names = {"--keep-wildcards"},
            description = "Keep sequences with wildcards in the output")
    public boolean keepWildcards = false;

    @Override
    protected List<String> getInputFiles() {
        return Arrays.asList(bamFiles);
    }

    @Override
    protected List<String> getOutputFiles() {
        return Stream.of(fastq1, fastq2, fastqUnpaired).filter(Objects::nonNull).collect(Collectors.toList());
    }

    @Override
    public void run0() throws Exception {
        try (BAMReader converter = new BAMReader(bamFiles, dropNonVDJ, !keepWildcards)) {
            SequenceRead read;
            try (PairedFastqWriter wr = new PairedFastqWriter(fastq1, fastq2);
                 SingleFastqWriter swr = new SingleFastqWriter(fastqUnpaired)) {
                while ((read = converter.take()) != null) {
                    if (read instanceof PairedRead) {
                        wr.write((PairedRead) read);
                    } else if (read instanceof SingleRead) {
                        swr.write((SingleRead) read);
                    }
                }
            }

            System.out.println("Your fastq files are ready.");
            System.out.println(converter.getNumberOfProcessedAlignments() + " alignments processed.");
            System.out.println(converter.getNumberOfPairedReads() + " paired reads.");
            System.out.println(converter.getNumberOfUnpairedReads() + " unpaired reads.");
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandBAM2fastq.kt
/*
 * Copyright (c) 2014-2023, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import cc.redberry.pipe.util.forEach
import com.milaboratory.app.logger
import com.milaboratory.core.io.sequence.PairedRead
import com.milaboratory.core.io.sequence.SingleRead
import com.milaboratory.core.io.sequence.fastq.PairedFastqWriter
import com.milaboratory.core.io.sequence.fastq.SingleFastqWriter
import com.milaboratory.mixcr.bam.BAMReader
import com.milaboratory.util.TempFileManager
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import java.nio.file.Path

@Command(
    name = "BAM2fastq",
    hidden = true,
    description = ["Converts BAM/SAM file to paired/unpaired fastq files"]
)
class CommandBAM2fastq : MiXCRCommandWithOutputs() {
    @Option(
        description = ["Put temporary files in the same folder as the output files."],
        names = ["--use-local-temp"],
        order = OptionsOrder.localTemp
    )
    var useLocalTemp = false

    @Option(
        names = ["-b", "--bam"],
        description = ["BAM files for conversion."],
        required = true,
        paramLabel = "<path>",
        order = OptionsOrder.main + 10_100
    )
    lateinit var bamFiles: List<Path>

    @Option(
        names = ["-r1"],
        description = ["File for first reads."],
        required = true,
        paramLabel = "<path>",
        order = OptionsOrder.main + 10_200
    )
    lateinit var fastq1: Path

    @Option(
        names = ["-r2"],
        description = ["File for second reads."],
        required = true,
        paramLabel = "<path>",
        order = OptionsOrder.main + 10_300
    )
    lateinit var fastq2: Path

    @Option(
        names = ["-u"],
        description = ["File for unpaired reads."],
        required = true,
        paramLabel = "<path>",
        order = OptionsOrder.main + 10_400
    )
    lateinit var fastqUnpaired: Path

    @Option(
        names = ["--drop-non-vdj"],
        description = ["Drop reads from bam file mapped on human chromosomes except with VDJ region (2, 7, 14, 22)"],
        order = OptionsOrder.main + 10_500
    )
    var dropNonVDJ = false

    @Option(
        names = ["--keep-wildcards"],
        description = ["Keep sequences with wildcards in the output"],
        order = OptionsOrder.main + 10_600
    )
    var keepWildcards = false

    @Option(
        names = [BAMReader.referenceForCramOption],
        description = ["Reference for genome that was used for build a cram file"],
        order = OptionsOrder.main + 10_700
    )
    var referenceForCram: Path? = null

    override val inputFiles
        get() = bamFiles.toList()

    override val outputFiles
        get() = listOf(fastq1, fastq2, fastqUnpaired)

    override fun run1() {
        val tempFileDest = TempFileManager.smartTempDestination(fastq1, "", !useLocalTemp)
        BAMReader(bamFiles, dropNonVDJ, !keepWildcards, tempFileDest, referenceForCram).use { converter ->
            PairedFastqWriter(fastq1.toFile(), fastq2.toFile()).use { wr ->
                SingleFastqWriter(fastqUnpaired.toFile()).use { swr ->
                    converter.forEach { read ->
                        when (read) {
                            is PairedRead -> wr.write(read)
                            is SingleRead -> swr.write(read)
                            else -> throw IllegalArgumentException()
                        }
                    }
                }
            }
            logger.log("Your fastq files are ready.")
            logger.log("${converter.numberOfProcessedAlignments()} alignments processed.")
            logger.log("${converter.numberOfPairedReads()} paired reads.")
            logger.log("${converter.numberOfUnpairedReads()} unpaired reads.")
        }
    }
}
